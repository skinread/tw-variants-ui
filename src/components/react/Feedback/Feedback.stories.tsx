import type { Meta, StoryObj } from '@storybook/react';

import { Feedback } from './Feedback';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: 'React/Feedback',
  component: Feedback,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  args: {
    children: 'This is error feedback content',
    color: 'error',
    iconName: 'exclamation',
  },
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Feedback>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Error: Story = {};

export const Success: Story = {
  args: {
    children: 'Your update has been saved.',
    color: 'success',
    iconName: 'tick-circle',
  },
};
