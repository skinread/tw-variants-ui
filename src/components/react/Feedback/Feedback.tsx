import { feedbackVariants, type FeedbackVariantsWithDocs } from '@common';
import { Icon, type FeedbackIcons } from '@components/react';

export interface FeedbackProps extends FeedbackVariantsWithDocs {
  /**
   * The feedback message content passthrough. Prefer text string or string array to JSX children
   */
  children: React.ReactNode | string[];
  /**
   * Icon to display
   */
  iconName?: FeedbackIcons;
  /**
   * Please remember to add an aria live attribute if the feedback is dynamically rendered
   */
  'aria-live'?: React.HTMLAttributes<HTMLDivElement>['aria-live'];
}

/**
 * Feedback component for consistent alert messages. If the feedback is shown conditionally,
 * please use the `aria-live` attribute to ensure screen readers announce the message.
 */
export const Feedback = (props: FeedbackProps) => {
  const { children, color, iconName = 'exclamation' } = props;
  const {
    container: containerStyle,
    icon: iconStyle,
    text: textStyle,
  } = feedbackVariants({ color });

  const Message = () => {
    if (typeof children === 'string') return <p>{children}</p>;
    // just a little custom logic to handle arrays of strings
    if (Array.isArray(children) && children.every((child) => typeof child === 'string')) {
      return children.map((text) => <p key={text.substring(0, 2)}>{text}</p>);
    }

    return children;
  };

  return (
    <div className={containerStyle()}>
      <Icon name={iconName} className={iconStyle()} />
      <div className={textStyle()}>
        <Message />
      </div>
    </div>
  );
};
