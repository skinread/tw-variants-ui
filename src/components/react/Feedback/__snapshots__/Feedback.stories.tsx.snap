// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Feedback Error smoke-test 1`] = `
<div class="alert p-4 border-0 border-l-[6px] rounded-md alert-error bg-red-light-3 border-error">
  <svg xmlns="http://www.w3.org/2000/svg"
       viewbox="0 0 24 24"
       fill="currentColor"
       class="inline-block  w-8 h-8 fill-error icon-exclamation"
       aria-hidden="true"
  >
    <path d="m22.242,5.272l-3.515-3.515c-1.133-1.133-2.64-1.757-4.242-1.757h-4.971c-1.603,0-3.109.624-4.242,1.757l-3.515,3.514c-1.134,1.133-1.758,2.64-1.758,4.243v4.971c0,1.603.624,3.11,1.758,4.243l3.515,3.515c1.133,1.133,2.64,1.757,4.242,1.757h4.971c1.603,0,3.109-.624,4.242-1.757l3.515-3.514c1.134-1.133,1.758-2.64,1.758-4.243v-4.971c0-1.603-.624-3.11-1.758-4.243Zm-.242,9.213c0,1.069-.416,2.073-1.172,2.829l-3.515,3.515c-.756.755-1.76,1.171-2.828,1.171h-4.971c-1.068,0-2.072-.416-2.828-1.171l-3.515-3.515c-.756-.755-1.172-1.759-1.172-2.828v-4.971c0-1.069.416-2.073,1.172-2.829l3.515-3.515c.756-.755,1.76-1.171,2.828-1.171h4.971c1.068,0,2.072.416,2.828,1.171l3.515,3.515c.756.755,1.172,1.759,1.172,2.828v4.971Zm-11-.485V6c0-.552.447-1,1-1s1,.448,1,1v8c0,.552-.447,1-1,1s-1-.448-1-1Zm2.5,3.5c0,.828-.672,1.5-1.5,1.5s-1.5-.672-1.5-1.5.672-1.5,1.5-1.5,1.5.672,1.5,1.5Z"
          stroke="none"
    >
    </path>
  </svg>
  <div class="text-foreground [overflow-wrap:anywhere] pt-1">
    <p>
      This is error feedback content
    </p>
  </div>
</div>
`;

exports[`React/Feedback Success smoke-test 1`] = `
<div class="alert p-4 border-0 border-l-[6px] rounded-md alert-success bg-green-light-3 border-success">
  <svg xmlns="http://www.w3.org/2000/svg"
       viewbox="0 0 24 24"
       fill="currentColor"
       class="inline-block  w-8 h-8 fill-success icon-tick-circle"
       aria-hidden="true"
  >
    <path d="m16.298,8.288l1.404,1.425-5.793,5.707c-.387.387-.896.58-1.407.58s-1.025-.195-1.416-.585l-2.782-2.696,1.393-1.437,2.793,2.707,5.809-5.701Zm7.702,3.712c0,6.617-5.383,12-12,12S0,18.617,0,12,5.383,0,12,0s12,5.383,12,12Zm-2,0c0-5.514-4.486-10-10-10S2,6.486,2,12s4.486,10,10,10,10-4.486,10-10Z"
          stroke="none"
    >
    </path>
  </svg>
  <div class="text-foreground [overflow-wrap:anywhere] pt-1">
    <p>
      Your update has been saved.
    </p>
  </div>
</div>
`;
